lib/pos_blockchain.go

```diff
@@ -119,12 +120,122 @@ func (bc *Blockchain) processOrphanBlockPoS(block *MsgDeSoBlock) error {
+	currentEpochEntry, err := utxoView.GetCurrentEpochEntry()
+	if err != nil {
+		// What do we want to do here? We can't validate the QC without getting the current epoch entry.
+		return errors.Wrap(err, "processOrphanBlockPoS: Problem getting current epoch entry")
+	}
+	var validatorsByStake []*ValidatorEntry
+	if !currentEpochEntry.ContainsBlockHeight(block.Header.Height) {
+		// Get the epoch entry based on the block height. The logic is the same
+		// regardless of whether the block is in a previous or future epoch.
+		// Note that the InitialView cannot be properly computed.
+		var epochEntry *EpochEntry
+		usePrevEpoch := block.Header.Height < currentEpochEntry.InitialBlockHeight
+		// If it's in a previous epoch, we compute the prev epoch entry.
+		if usePrevEpoch {
+			epochEntry, err = utxoView.computePrevEpochEntry(currentEpochEntry.EpochNumber, currentEpochEntry.InitialBlockHeight, currentEpochEntry.InitialView, 0)
+			if err != nil {
+				return errors.Wrap(err, "processOrphanBlockPoS: Problem computing prev epoch entry")
+			}
+		} else {
+			// Okay now we know that this block must be in a future epoch. We do our best to compute
+			// the next epoch entry and check if its in that epoch. If it's in a future epoch, we just throw it away.
+			// We supply 0 for the view and 0 for the block timestamp as we don't know what those values should be and
+			// we will ignore these values.
+			epochEntry, err = utxoView.computeNextEpochEntry(currentEpochEntry.EpochNumber, currentEpochEntry.FinalBlockHeight, 0, 0)
+			if err != nil {
+				return errors.Wrap(err, "processOrphanBlockPoS: Problem computing next epoch entry")
+			}
+		}
+		if !epochEntry.ContainsBlockHeight(block.Header.Height) {
+			// We will throw away this block as we know it's not in either
+			// the next or the previous epoch.
+			errSuffix := "future"
+			if usePrevEpoch {
+				errSuffix = "past"
+			}
+			return fmt.Errorf("processOrphanBlockPoS: Block height %d is too far in the %v", block.Header.Height, errSuffix)
+		}
+		var epochEntrySnapshotAtEpochNumber uint64
+		epochEntrySnapshotAtEpochNumber, err = utxoView.ComputeSnapshotEpochNumberForEpoch(epochEntry.EpochNumber)
+		if err != nil {
+			return errors.Wrapf(err, "processOrphanBlockPoS: Problem getting snapshot at epoch number for poch entry at epoch #%d", epochEntry.EpochNumber)
+		}
+		// Okay now that we've gotten the SnapshotAtEpochNumber for the next epoch, we can make sure that the proposer
+		// of the block is within the set of potential block proposers for the next epoch.
+		snapshotLeaderPKIDs, err := utxoView.GetSnapshotLeaderScheduleAtEpochNumber(epochEntrySnapshotAtEpochNumber)
+		if err != nil {
+			return errors.Wrapf(err, "processOrphanBlockPoS: Problem getting snapshot leader schedule at snapshot at epoch number %d", epochEntrySnapshotAtEpochNumber)
+		}
+		// Get the PKID for the block proposer.
+		blockProposerPKID := utxoView.GetPKIDForPublicKey(block.Header.ProposerPublicKey.ToBytes()).PKID
```

Comment:
I guess we run the risk here of the PKID <-> Public key mapping having changed since the leader schedule was snapshotted

Reply:
Temporarily resolved offline. We'll run with this check for now and iterate on it later